### PR TITLE
Reduce Tech dept - Code coverage and quality 

### PR DESCRIPTION
**Pull request by ronenhamias**
add example test working with cluster configuration for advanced settings.

 Microservices node = Microservices.builder().clusterConfig(
        ClusterConfig.builder()

            .transportConfig(TransportConfig.builder()
                .port(3000)
                .portAutoIncrement(false)
                .build())) // cluster config
        
        .build(); // node

### CODE REVIEW

File: services/src/main/java/io/scalecube/services/ServiceHeaders.java [hunk 1]
```diff
@@ -30,40 +30,36 @@
    * header is used to mark a request as a service response.
    */
   public static final String SERVICE_RESPONSE = "service-response";
-
-  private ServiceHeaders() {
-    // Do not instantiate
```
Comment (antonkharenko) [inline services/src/main/java/io/scalecube/services/ServiceHeaders.java]: Why did you remove this constructor it seems you don't want to instantiate this class?

File: services/src/test/java/io/scalecube/services/DummyRouter.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  private Object empty;
+
+  public DummyRouter() {
+    checkArgument(this.empty != null);
+  }
+
+  @Override
+  public Optional<ServiceInstance> route(Message request) {
+    // TODO Auto-generated method stub
+    return null;
```
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/DummyRouter.java]: Basically if `null` is supposed behavior then better to return empty optional instead and remove TODO comment above...

File: services/src/main/java/io/scalecube/services/ServiceHeaders.java (hunk 1 above)
Comment (ronenhamias) [inline services/src/main/java/io/scalecube/services/ServiceHeaders.java]: nope just its all static and code cheker does not like it

File: services/src/test/java/io/scalecube/services/GreetingServiceImpl.java [hunk 3]
```diff
@@ -9,6 +9,11 @@
 final class GreetingServiceImpl implements GreetingService {
 
   @Override
+  public String toString() {
+    return "GreetingServiceImpl []";
```
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/GreetingServiceImpl.java]: Why do you need toString for service? Also default implementation of toString will be about the same informative

File: services/src/test/java/io/scalecube/services/DummyRouter.java (hunk 2 above)
Comment (ronenhamias) [inline services/src/test/java/io/scalecube/services/DummyRouter.java]: this dummy router throw runtime exception when creating it 
this is for covering the case of excpetion when creating router for code coverage

File: services/src/test/java/io/scalecube/services/GreetingServiceImpl.java (hunk 3 above)
Comment (ronenhamias) [inline services/src/test/java/io/scalecube/services/GreetingServiceImpl.java]: since i need expected behavior for tests

File: services/src/test/java/io/scalecube/services/ServiceConfigTest.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        ClusterConfig.builder()
+
+            .transportConfig(TransportConfig.builder()
+                .port(3000)
+                .portAutoIncrement(false)
+                .build())) // cluster config
+        
+        .build(); // node 
+
+    assertTrue(node.cluster().address().port() == 3000);
```
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/ServiceConfigTest.java]: you need to shutdown node.

File: services/src/test/java/io/scalecube/services/ServiceUnitTest.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import io.scalecube.transport.Address;
+import io.scalecube.transport.Message;
+
+import org.junit.Test;
+
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Optional;
+
+public class ServiceUnitTest {
```
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/ServiceUnitTest.java]: ServiceUnitTest seems to be an odd name

File: services/src/test/java/io/scalecube/services/ServiceUnitTest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    Message request = Message.builder().correlationId(response.correlationId()).build();
+    Field field = Message.class.getDeclaredField("sender");
+    field.setAccessible(true);
+    field.set(request, member.cluster().address());
+
+    DispatchingFuture dispatcher = DispatchingFuture.from(member.cluster(), request);
+    dispatcher.complete(new Throwable());
+
+    assertTrue(true);
```
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/ServiceUnitTest.java]: this is seems to be a bit strange check, are you really need it )

File: services/src/main/java/io/scalecube/services/LocalServiceInstance.java [hunk 7]
```diff
@@ -55,11 +55,7 @@ public Object invoke(Message message) throws Exception {
       if (method.getParameters().length == 0) {
         result = method.invoke(serviceObject);
       } else if (method.getParameters()[0].getType().isAssignableFrom(Message.class)) {
-        if (message.data().getClass().isAssignableFrom(Message.class)) {
-          result = method.invoke(serviceObject, (Message) message.data());
-        } else {
-          result = method.invoke(serviceObject, message);
-        }
+        result = method.invoke(serviceObject, (Message) message.data());
```
Comment (antonkharenko) [inline services/src/main/java/io/scalecube/services/LocalServiceInstance.java]: Not sure that I know the logic supposed to be here, but is it really that message.data() is always an Message here?

File: services/src/test/java/io/scalecube/services/ServiceUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    assertTrue(instance.toString().equals("LocalServiceInstance [serviceObject=GreetingServiceImpl [], memberId=a]"));
+    assertTrue(instance.tags().isEmpty());
+    assertTrue(instance.memberId().equals("a"));
+    assertTrue(instance.serviceName().equals("b"));
+  }
+
+  @Test
+  public void test_remote_service_instance() {
+
+    Microservices member = Microservices.builder().build();
```
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/ServiceUnitTest.java]: I think you need to shutdown this member at the end.

File: services/src/main/java/io/scalecube/services/ServiceHeaders.java (hunk 1 above)
Comment (antonkharenko) [inline services/src/main/java/io/scalecube/services/ServiceHeaders.java]: but the reason to define private default constructor is to forbid anyone to instantiate this class.

File: services/src/test/java/io/scalecube/services/DummyRouter.java (hunk 2 above)
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/DummyRouter.java]: still remove TODO comment

File: services/src/test/java/io/scalecube/services/GreetingServiceImpl.java (hunk 3 above)
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/GreetingServiceImpl.java]: are you testing it somehow via toString? in such case maybe better to test in some more straight forward way.

File: services/src/main/java/io/scalecube/services/ServiceHeaders.java (hunk 1 above)
Comment (ronenhamias) [inline services/src/main/java/io/scalecube/services/ServiceHeaders.java]: yes but automatic code review does not like it

File: services/src/test/java/io/scalecube/services/ServiceUnitTest.java (hunk 5 above)
Comment (ronenhamias) [inline services/src/test/java/io/scalecube/services/ServiceUnitTest.java]: what is the convention?

File: services/src/test/java/io/scalecube/services/ServiceUnitTest.java (hunk 6 above)
Comment (ronenhamias) [inline services/src/test/java/io/scalecube/services/ServiceUnitTest.java]: this testing dispatcher with throwable

File: services/src/test/java/io/scalecube/services/GreetingServiceImpl.java (hunk 3 above)
Comment (ronenhamias) [inline services/src/test/java/io/scalecube/services/GreetingServiceImpl.java]: the test test the to string
that the expected string returns

File: services/src/main/java/io/scalecube/services/ServiceHeaders.java (hunk 1 above)
Comment (antonkharenko) [inline services/src/main/java/io/scalecube/services/ServiceHeaders.java]: There are no issues related to it.

File: services/src/test/java/io/scalecube/services/ServiceUnitTest.java (hunk 5 above)
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/ServiceUnitTest.java]: Typically for unit tests the convention is `<Class name you test>Test`, also I see in this case there are several tests not really related to each other. So maybe create several test classes where each will correspond to specific class you test or give some meaningful name to the test to display the fact why those tests are collocated in single file and remove word `Unit` from the name.

File: services/src/test/java/io/scalecube/services/ServiceUnitTest.java (hunk 6 above)
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/ServiceUnitTest.java]: yes, but since the only assert you have here is `assertTrue(true)` which obviously passing no matter what you will do above it is kind of allude that it didn't really test anything.

File: services/src/test/java/io/scalecube/services/GreetingServiceImpl.java (hunk 3 above)
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/GreetingServiceImpl.java]: Which one is testing this toString?

File: services/src/test/java/io/scalecube/services/ServiceUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    ServiceRegistry registry = new ServiceRegistryImpl(member.cluster(), config, processor);
+    RemoteServiceInstance instance = new RemoteServiceInstance(registry, reference, new HashMap<>());
+
+    assertTrue(instance.toString().equals("RemoteServiceInstance [address=localhost:4000, memberId=a]"));
+    assertTrue(instance.tags().isEmpty());
+    assertTrue(instance.memberId().equals("a"));
+    assertTrue(instance.address().equals(Address.create("localhost", 4000)));
+    assertTrue(!instance.isReachable());
+    assertTrue(!instance.isLocal());
+    assertTrue(instance.serviceName().equals("b"));
```
Comment (antonkharenko) [inline services/src/test/java/io/scalecube/services/ServiceUnitTest.java]: I think for such cases it is better to use `assertEquals` instead of `assertTrue` - it will give much more helpful error message in case if assert won't pass.

**Comment by antonkharenko** [comment]
@ronenhamias due to some reasons after your commit it shows many of my comments as outdated discussions even so they wasn't addressed at any way... so please take a look also to outdated discussions since it doesn't make sense to write them second time.

**Comment by ronenhamias** [comment]
sure i will review them

**Comment by ronenhamias** [comment]
@antonkharenko i have related to all your comments in my last commit
i cannot reply anymore in the specific comment due to this outdated stuff
